---
title: "assessment3"
author: "Lahirunee Rupasinghe & Tayla Gibson"
date: '2022-05-13'
output:
  html_document: default
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

## R Markdown

This is an R Markdown document. Markdown is a simple formatting syntax for authoring HTML, PDF, and MS Word documents. For more details on using R Markdown see <http://rmarkdown.rstudio.com>.

When you click the **Knit** button a document will be generated that includes both content as well as the output of any embedded R code chunks within the document. You can embed an R code chunk like this:

```{r cars}
summary(cars)
```

## Including Plots

You can also embed plots, for example:

```{r pressure, echo=FALSE}
plot(pressure)
```

Note that the `echo = FALSE` parameter was added to the code chunk to prevent printing of the R code that generated the plot.


## Part 1: Gene expression

This is my introduction to analysing gene expression.

```{r pip}
# Read in data file

URL="https://raw.githubusercontent.com/markziemann/SLE712_files/master/assessment_task3/bioinfo_asst3_part1_files/gene_expression.tsv"

pip <- read.table(URL,header = TRUE,row.names = 1)

head(pip)
```


## Part 1: Growth data

This is my introduction to analysing of growth data.

```{r mychunk}
# Read in data file

## GROWTH DATA FILE

### How to download file


download.file(url = "https://raw.githubusercontent.com/markziemann/SLE712_files/master/assessment_task3/bioinfo_asst3_part1_files/growth_data.csv", 
              destfile = "growth_data.csv")

### Reading file into RSTudio
d <- read.csv("growth_data.csv")
head(d)
### To check structure of data 
str(d)
### To deactivate the default setting of recognising strings as factors and remove extra headings
d <- read.csv("growth_data.csv", stringsAsFactors = FALSE, header = TRUE)
head(d)
str(d)


###Question 6

#### To find column names
colnames(d, do.NULL = TRUE, prefix = "col")



### Question 7 

#### To calculate mean
##### To do row-wise subsetting for Site=="southwest"
d_sw <- d[d$Site=="southwest", ]
head(d_sw)
##### To do row-wise subsetting for Site=="northeast"
d_ne <- d[d$Site=="northeast", ]
head(d_ne)

#### To calculate standard deviation
##### To calculate standard deviation for Site=="northeast" 
ne<-list()
ne$mean<-colMeans(d_ne[,c(3,6)])
ne$std<- c(sd(d_ne[,3]), sd(d_ne[,6]))
##### To present standard deviation for Site=="northeast"  
ne <- as.data.frame(ne)
head(ne)
##### To calculate standard deviation for Site=="southwest"
sw<-list()
sw$mean<-colMeans(d_sw[,c(3,6)])
sw$sd<- c(sd(d_sw[,3]), sd(d_sw[,6]))
##### To present standard deviation for Site=="southwest" 
sw <- as.data.frame(sw);
head(sw)

### Question 8

#### Making a boxplot
ne <- subset(d, Site=="northeast")
head(ne)
sw <- subset(d, Site=="southwest")
head(sw)
```


## Part 2: Examining Biological Sequence Diversity

Comparing E. coli to Mycoplasma hyopneumoniae str. Mycoplasma hyopneumoniae.

```{r sequence data}
# Loading sequence data

library("seqinr")
library("kableExtra")
library("R.utils")

URL="http://ftp.ensemblgenomes.org/pub/release-53/bacteria//fasta/bacteria_0_collection/escherichia_coli_str_k_12_substr_mg1655_gca_000005845/cdna/Escherichia_coli_str_k_12_substr_mg1655_gca_000005845.ASM584v2.cdna.all.fa.gz"
download.file(URL,destfile="ecoli_cds.fa.gz")
gunzip("ecoli_cds.fa.gz")
list.files()

cds <- seqinr::read.fasta("ecoli_cds.fa")
str(head(cds))

URL="http://ftp.ensemblgenomes.org/pub/release-53/bacteria//fasta/bacteria_40_collection/mycoplasma_hyopneumoniae_gca_004768725/cdna/Mycoplasma_hyopneumoniae_gca_004768725.ASM476872v1.cdna.all.fa.gz"
download.file(URL,destfile="mhp_cds.fa.gz")
gunzip("mhp_cds.fa.gz")
list.files()

cds <- seqinr::read.fasta("mhp_cds.fa")
str(head(cds))
```


Note that the `echo = FALSE` parameter was added to the code chunk to prevent printing of the R code that generated the plot.

# Lahi was here

