---
title: "Part 1"
author: "Lahirunee Rupasinghe & Tayla Gibson"
date: '2022-05-27'
output: html_document
theme: yeti
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

## Part 1: Importing files, data wrangling, mathematical operations, plots and saving code on GitHub

This part of Assessment 3 involves working with the data files for [Gene expression](https://github.com/markziemann/SLE712_files/blob/master/assessment_task3/bioinfo_asst3_part1_files/gene_expression.tsv) and [Growth data](https://github.com/markziemann/SLE712_files/blob/master/assessment_task3/bioinfo_asst3_part1_files/growth_data.csv).

Source: https://github.com/tlgibson21/assessment3/blob/main/part1.Rmd

## Section 1: Gene expression

This section will demonstrate how to analyse gene expression with RNA sequencing count data and collate the data into a histogram plot.

```{r kable, echo=FALSE}
suppressPackageStartupMessages({
library("kableExtra")
})
```

**1. Read in the file, making the gene identifiers the row names. Show a table of values for the first six genes.**

To solve this problem, I opened [Gene expression](https://github.com/markziemann/SLE712_files/tree/master/assessment_task3/bioinfo_asst3_part1_files) and clicked on [View Raw](https://raw.githubusercontent.com/markziemann/SLE712_files/master/assessment_task3/bioinfo_asst3_part1_files/gene_expression.tsv).

I then copied the URL and used the `read.table()` command to read in the data file.

I added `header=TRUE` to remove the variable numbers at the start of each column.

I then added `rownames=1` to make the gene identifiers the row names. 

Next, I used the assignment operator `<-` to name the function `gene_exp`.

The `head()` function allowed me to display the first six genes, and the kable code styled my table.

```{r gene_exp}
# Read in data file

URL="https://raw.githubusercontent.com/markziemann/SLE712_files/master/assessment_task3/bioinfo_asst3_part1_files/gene_expression.tsv"

gene_exp <- read.table(URL, header=TRUE, row.names=1)

dt1 <- head(gene_exp) # show the first part

dt1 %>%
  kbl(caption = "Table 1: Gene Expression Data Table") %>%
  kable_styling(full_width = F, html_font = "Arial") # create kableExtra table
```

**2. Make a new column which is the mean of the other columns. Show a table of values for the first six genes.**

To create a new column, I used the `$` notation in between the `gene_exp` data set and the new column name **Mean**.

The `apply()` function was used to gather the mean, where `gene_exp` represented the matrix, `1` indicates the rows and `mean` represents the function being applied.

Once more, the `head()` function was employed to demonstrate the first six genes and kable styling was executed.

```{r mean}
gene_exp$Mean <- apply(gene_exp, 1, mean) # create a new column

dt2 <- head(gene_exp) # show the first six genes

dt2 %>%
  kbl(caption = "Table 2: Gene Expression Mean Data") %>%
  kable_styling(full_width = F, html_font = "Arial") # create kableExtra table
```

**3. List the 10 genes with the highest mean expression.**

Firstly, I sorted the mean expressions using the `order()` function, which I encompassed within a set of square brackets as they can select more than one element from the data.

In this scenario, I sorted the **Mean** column, and used the minus sign `-` to reverse the order so that the highest means are shown first in the list.

I then used  `<-` to name the function `sorted`. That could then be applied in the `head()` function specifying 10 genes by adding 10 into the function.

I then named this function `list` so that I could designate just the gene names, by using `rownames()`, and applying it to the list I had created.

Lastly, I created another table with the kableExtra package.

```{r ten}
sorted <- gene_exp[order(-gene_exp$Mean),] # sort the mean

list <- head(sorted, 10) # list 10

dt3 <- rownames(list) # define the row names

dt3 %>%
  kbl(caption = "Table 3: Genes with the Highest Mean Expression") %>%
  kable_styling(full_width = F, html_font = "Arial") # create kableExtra table
```

**4. Determine the number of genes with a mean < 10**

To gather these genes, I used the `subset` function to set the parameters as `<10` for the **Mean** column in the **Gene expression** data set.

I defined this function as `less than` using `<-`.

I was then able to apply that variable within the `nrow()` function to determine how many rows, and therefore, how many genes are defined by that parameter.

```{r below}
lessthan <- subset(gene_exp,Mean < 10) # subset the mean less than 10

nrow(lessthan) # show the row names of this selection
```

**5. Make a histogram plot of the mean values and include it into your report.**

The `hist()` function was used to make a histogram plot of the mean values gathered by `gene_exp$Mean`.

I included `xlab=` to add a title for the x-axis and `main=` to add a title for the histogram plot.

```{r hist}
# Plot a histogram

hist(gene_exp$Mean, xlab="Mean",
     main="Histogram of Mean Values")
```

## Section 2: Growth data

This section will demonstrate how to import and analyse growth data.

**6. Read in the growth data file, import into an  R object, calculate mean and standard deviation of tree circumference, make a box plot and estimate the p-value.**

To solve this problem, I opened [Growth data](https://github.com/markziemann/SLE712_files/tree/master/assessment_task3/bioinfo_asst3_part1_files) and clicked on **Raw**. I then copied the URL and used the `growth_d <- read.csv("growth_data.csv")`command to read in the data file. Next, I used the assignment operator `<-` to name the function `grwoth_d`.I added `header=TRUE`  to deactivate the default setting of recognising strings as factors and remove extra headings. Then I added `str(growth_d)` to check the structure of data.

```{r growth_data}
# How to download file

download.file(url = "https://raw.githubusercontent.com/markziemann/SLE712_files/master/assessment_task3/bioinfo_asst3_part1_files/growth_data.csv", 
              destfile = "growth_data.csv")

# Reading file into RSTudio

growth_d <- read.csv("growth_data.csv")

head(growth_d)

# To check structure of data

str(growth_d)

# To deactivate the default setting of recognising strings as factors and remove extra headings

growth_d <- read.csv("growth_data.csv", stringsAsFactors = FALSE, header = TRUE)

head(growth_d)

str(growth_d)

# To find column names

colnames(growth_d, do.NULL = TRUE, prefix = "col")
```

**7. Calculate the mean and standard deviation of tree circumference at the start and end of the study at both sites.**

To solve this problem, I subset the each site `site1 <- subset(growth_d,Site == "northeast")` and `site2 <- subset(growth_d,Site == "southwest")`.

To find the mean and standard deviation in northeast site, I added `start_site1 <- site1$Circumf_2005_cm` then`mean(start_site1)` and `sd(start_site1)` and will be repeated for the southwest site.

```{r mean and sd}
# Subset each site

site1 <- subset(growth_d,Site == "northeast")

site2 <- subset(growth_d,Site == "southwest")

# Mean and standard deviation at the start of the study at the first site 'northeast'

start_site1 <- site1$Circumf_2005_cm

mean(start_site1)

sd(start_site1)

# Mean and standard deviation at the end of the study at the first site 'northeast'

end_site1 <- site1$Circumf_2020_cm

mean(end_site1)

sd(end_site1)

# Mean and standard deviation at the start of the study at the second site 'southwest'

start_site2 <- site2$Circumf_2005_cm

mean(start_site2)

sd(start_site2)

# Mean and standard deviation at the end of the study at the second site 'southwest'

end_site2 <- site2$Circumf_2020_cm

mean(end_site2)

sd(end_site2)
```

**8. Make a box plot of tree circumference at the start and end of the study at both sites.**

To make a box plot, I added, the below coding and I added, `main="Tree Circumference at the Start and End of the study at Both Sites"` to make the header of the plot and then I added ` ylab="Tree Circumference (cm)")`to label the Y-axis of plot. Finally, I added `grid()` to add grid to the plot.

```{r Box plot}
# How to make a box plot

boxplot(site1$Circumf_2005_cm,site1$Circumf_2020_cm,
        site2$Circumf_2005_cm,site2$Circumf_2020_cm,
        names=c("2005 at NE site","2020 at NE site",
                "2005 at SW site","2020 at SW site"),
        main="Tree Circumference at the Start and End of the study at Both Sites",
        ylab="Tree Circumference (cm)")

grid()
```

**9. Calculate the mean growth over the last 10 years at each site.**

To solve this problem, I subset each site separately using `mid_site1 <- site1$Circumf_2010_cm` and `mid_site2 <- site2$Circumf_2010_cm`.

Then, I added the coding for growth rate for each site using `growth_rate_site1 <- ((end_site1-mid_site1)/mid_site1)` and `growth_rate_site2 <- ((end_site2-mid_site2)/mid_site2)`

```{r Mean for the last 10 years}
# How to calculate the mean growth over the last 10 years

mid_site1 <- site1$Circumf_2010_cm
mid_site2 <- site2$Circumf_2010_cm

growth_rate_site1 <- ((end_site1-mid_site1)/mid_site1)
growth_rate_site2 <- ((end_site2-mid_site2)/mid_site2)

mean_site1 <- mean(growth_rate_site1)

mean_site2 <- mean(growth_rate_site2)

mean_growth <- mean(growth_rate_site1+growth_rate_site2)
```

**10. Use the t.test and wilcox.test functions to estimate the p-value that the 10 year growth is different at the two sites.**

To solve this problem, I added a function `res <- t.test (growth_rate_site1,growth_rate_site2)` and then `p.value <- res$p.value` to find the p value.

To round the number into 3 digits I added `HEADER <- paste("P Value:" , signif(p.value,3))`.

```{r t test}
res <- t.test (growth_rate_site1,growth_rate_site2)

p.value <- res$p.value

paste("P Value from T Test:" , signif(p.value,3))

res <- wilcox.test(growth_rate_site1,growth_rate_site2)

p.value2 <- res$p.value

paste("P Value from Wilcox Test:" , signif(p.value2,3))

###fkfdmbkfb
```
