# Assessment3

The GitHub repository **assessment3** is composed of two parts.
* part1.rmd
* part2.rmd

Part 1 contains a data set for gene expression which includes RNA sequencing count data for three samples of interest.
This section will demonstrate how to import the data and analyse the data.
Lastly, it will exhibit how to plot a histogram plot to observe the data generated.

Part 1 further contains a data set for growth data which includes the measurements for tree circumference of trees growing at two separate locations.
This section will demonstrate how to plot a box plot to visualise the differences at each location.

Part 2 uses coding DNA sequences from two organisms, E.coli and M. hyopneumoniae.
This part will demonstrate how to download and unzip sequence files, analyse the sequence data, evaluate nucleotide and amino acid frequency and demonstrate them using bar plots, estimate codon usage bias and evaluate protein sequence k-mer frequencies.

## How to run the programs?

The RMarkdown files can be run using RStudio. It may help to have the [*E.coli*](http://ftp.ensemblgenomes.org/pub/release-53/bacteria//fasta/bacteria_0_collection/escherichia_coli_str_k_12_substr_mg1655_gca_000005845/cdna/) and [*M. hyopneumoniae*](http://ftp.ensemblgenomes.org/pub/release-53/bacteria//fasta/bacteria_40_collection/mycoplasma_hyopneumoniae_gca_004768725/cdna/) cds files downloaded on your desktop using the links provided from [Ensembl Bacteria](https://bacteria.ensembl.org/index.html).
